modules/transaction-payment/src/lib.rs

```diff
@@ -611,60 +675,203 @@ where
-								.saturating_mul_int(target_price.saturating_mul_int(amount))
-						} else {
-							PalletBalanceOf::<T>::max_value()
-						};
-
-						if T::DEX::swap_with_exact_target(
-							who,
-							&trading_path,
-							amount.unique_saturated_into(),
-							<T as Config>::MultiCurrency::free_balance(supply_currency_id, who)
-								.min(max_supply_limit.unique_saturated_into()),
-						)
-						.is_ok()
-						{
-							// successfully swap, break iteration
-							break;
-						}
+	/// Iterate order list, break if can swap out enough native asset amount with user's foreign
+	/// asset. make sure trading path is exist in dex, if the trading pair is not exist in dex, even
+	/// though we have setup it in charge fee pool, we can't charge fee with this foreign asset.
+	fn swap_native_asset(who: &T::AccountId, amount: Balance) {
+		let native_currency_id = T::NativeCurrencyId::get();
+		for trading_path in Self::get_trading_path(who) {
+			if let Some(target_currency_id) = trading_path.last() {
+				if *target_currency_id == native_currency_id {
+					let supply_currency_id = *trading_path.first().expect("should match a non native asset");
+					if Self::swap_from_treasury_or_dex(who, amount, supply_currency_id).is_ok() {
+						break;
 					}
-					_ => {}
 				}
 			}
 		}
 	}
+
+	/// swap treasury account's native asset with user's foreign asset from treasury fee pool when
+	/// treasury account's native asset balance is gt threshold. or-else, swap treasury account's
+	/// native asset with treasury account's foreign asset from dex swap liquid pool.
+	#[transactional]
+	fn swap_from_treasury_or_dex(
+		who: &T::AccountId,
+		amount: Balance,
+		supply_currency_id: CurrencyId,
+	) -> DispatchResult {
+		let rate = TokenFixedRate::<T>::get(supply_currency_id).ok_or(Error::<T>::InvalidRate)?;
+		let treasury_account = Self::treasury_sub_account_id(supply_currency_id);
+
+		// if treasury account has not enough native asset, trigger swap from dex. if treasury_balance
+		// is lt ED, it become 0, this means the charge fee pool is exhausted. for prevent this case,
+		// we normally set the SwapBalanceThreshold gt ED. As we still use the old rate to calculate
+		// fee in this transaction. so for fee charge as successful as possible, we reserve one block
+		// fee, that's SwapBalanceThreshold can set to ED + one block fee.
+		let treasury_balance = T::Currency::free_balance(&treasury_account);
+		if treasury_balance < SwapBalanceThreshold::<T>::get(supply_currency_id) {
+			let trading_path = Self::get_trading_path_by_currency(&treasury_account, supply_currency_id);
+			if let Some(trading_path) = trading_path {
+				// using all supply asset to swap out native asset, no need to consider ED limitation here.
+				let treasury_supply_balance = T::MultiCurrency::free_balance(supply_currency_id, &treasury_account);
+				if let Ok(swap_native_balance) =
+					T::DEX::swap_with_exact_supply(&treasury_account, &trading_path, treasury_supply_balance, 0)
```

Comment:
although we use whitelist treasury sub account to store foreign asset from user, but when swap happen, if the foreign asset balance of treasury sub account still less than foreign asset's ED, the swap will failed and finally cause native balance really exhausted then charge fee pool will not works any more. we may get balance from treasury account like initial_pool phase when swap failed in this scene. I'll calcuate some data, and if need, add this logic later.

Reply:
the sub accounts are not whitelisted and I don't really want to add them to the ED whitelist
so I think the easy solution is ensure there is always ED in the sub accounts.
this means for all the transfer, query balance etc, it need to ensure there is always ED left
and in the initialization code, transfer ED amount from treasury account to sub account

Reply:
we really need to make sure the swap don't fail, and add a `else { debug_assert!(false, "Swap tx fee pool should not fail"); } ` to check it